src/Mod/Sketcher/App/planegcs/GCS.cpp

```diff
@@ -619,7 +618,7 @@ int System::addConstraint(Constraint *constr)
     if (constr->getTag() >= 0) // negatively tagged constraints have no impact
         hasDiagnosis = false;  // on the diagnosis
 
-    clist.push_back(constr);
+    clist.emplace_back(constr);
```

Comment:
So this is taking a pointer to a constraint constructed elsewhere, and turning it into a unique_ptr, which will then deallocate it? I don't think it's reasonable to go from a non-owning pointer and simply declare that you own it now: where is this function getting called?

Reply:
I couldn't agree more. :-)

But the important fact is that the `GCS::System` is actually in charge of destroying this `Constraint` someone else has allocated. :-(
Destructor:
https://github.com/FreeCAD/FreeCAD/blob/1dcc58c4b1c51c50215639ca58326f6b69867743/src/Mod/Sketcher/App/planegcs/GCS.cpp#L568
Called by destructor (Clear):
https://github.com/FreeCAD/FreeCAD/blob/1dcc58c4b1c51c50215639ca58326f6b69867743/src/Mod/Sketcher/App/planegcs/GCS.cpp#L590
Called by "Clear" (free):
https://github.com/FreeCAD/FreeCAD/blob/1dcc58c4b1c51c50215639ca58326f6b69867743/src/Mod/Sketcher/App/planegcs/GCS.cpp#L5588

So, the `GCS::System` actually owns what is passed to it through `addConstraint` and stored in `clist`.

To follow recommendations, we should actually change `addConstraint` to take a `unique_ptr`.

Reply:
> where is this function getting called?

Sorry... I hadn't answered this!
It is called all over the same .cpp file. There are different addXXXXConstraints:
https://github.com/FreeCAD/FreeCAD/blob/1dcc58c4b1c51c50215639ca58326f6b69867743/src/Mod/Sketcher/App/planegcs/GCS.cpp#L662

They do create a new constraint and "forget about it". Well... if `addConstraint` was "private" it would be forgivable. :-)